README.md
# grunt-phantomcss

> Automate CSS regression testing with PhantomCSS

## Notice

**This is a fork of the original (presumably discontinued). Currently this version here is untagged and unreleased on npm. You can install the original with "npm install grunt-phantomcss". However, you can install and use this version:**

Add this to your `package.json`:

    "grunt-phantomcss": "git://github.com/EmilioAiolfi/grunt-phantomcss.git",

or, alternatively, type this into your command line interface:

```shell
npm install --save-dev git://github.com/EmilioAiolfi/grunt-phantomcss.git
```
## Getting Started
This plugin requires Grunt `~0.4.5`

If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```javascript
grunt.loadNpmTasks('grunt-phantomcss');
```

## The "phantomcss" task

### Overview
In your project's Gruntfile, add a section named `phantomcss` to the data object passed into `grunt.initConfig()`.

```javascript
grunt.initConfig({
  phantomcss: {
    options: {},
    your_target: {
      options: {
        screenshots: 'test/visual/screenshots/',
        results: 'results/visual/',
        viewportSize: [1280, 800],
        mismatchTolerance: 0.05
      },
      src: [
        'test/visual/**/*.js'
      ]
    }
  }
});


grunt.registerTask('phantomcsstests', ['phantomcss']);
```


```shell
grunt phantomcsstests
```


### Options

#### baseURL - Optional
Type: `String`
Default: blank

Host for test - Use in test files: ```phantomcss.baseURL```

#### src
Type: `String|Array`

The test files to run.

#### options.mismatchTolerance
Type: `Number`
Default: `0.05`

Toleranz of errors that is allowed in a screenshot (for instance to match anti-aliasing bugs).

#### options.screenshots
Type: `String`
Default: `'./screenshots'`

The screenshots directory where test fixtures (comparison screenshots) are stored. Baseline screenshots will be stored here on the first run if they're not present.

#### options.results
Type: `String`
Default: `'./results'`

The directory to store source, diff, and failure screenshots after tests.

#### options.viewportSize
Type: `Array`
Default: `[1280, 800]`

The viewport size to test the site in `[width, height]` format. Useful when testing responsive layouts.

#### options.logLevel
Type: `String`
Default: `error`

The CasperJS log level. See [CasperJS: Logging](http://casperjs.readthedocs.org/en/latest/logging.html) for details.


### Usage Examples

#### Basic visual tests
Run tests in `test/visual/` against comparison screenshots stored in `test/visual/screenshots/`, and put the resulting screenshots in `results/visual/`

```javascript
grunt.initConfig({
  phantomcss: {
    options: {
      screenshots: 'test/visual/screenshots/',
      results: 'results/visual/'
    },
    src: [
      'test/visual/**/*.js'
    ]
  }
});
```

#### Responsive layout testing
Run tests in `test/visual/` against comparison screenshots for destop and mobile.

```javascript
grunt.initConfig({
  phantomcss: {
    desktop: {
      options: {
        screenshots: 'test/visual/desktop/',
        results: 'results/visual/desktop',
        viewportSize: [1024, 768]
      },
      src: [
        'test/visual/**.js'
      ]
    },
    mobile: {
      options: {
        screenshots: 'test/visual/mobile/',
        results: 'results/visual/mobile',
        viewportSize: [320, 480]
      },
      src: [
        'test/visual/**.js'
      ]
    }
  },
});
```

#### Sample test file

Test files should do the following:
* You no longer need ```casper.start``` - Use ```thenOpen``` to start CasperJS with the URL you want to test - NOT USE START (casper.start is default)
* Manipulate the page in some way
* Take screenshots

```javascript
casper.thenOpen('http://localhost:3000/')
.then(function() {
  phantomcss.screenshot('#todo-app', 'Main app');
})
.then(function() {
  casper.fill('form.todo-form', {
    todo: 'Item1'
  }, true);

  phantomcss.screenshot('#todo-app', 'Item added');
})
.then(function() {
  casper.click('.todo-done');

  phantomcss.screenshot('#todo-app', 'Item checked off');
});
```

You can also load a local file by specifying a path (relative to the Gruntfile):

```javascript
casper.thenOpen('build/client/index.html')
.then(function() {
  // ...
});
```

### Multiple Test Files
you no longer need ```casper.start```

File 1 - file1.js
```javascript
casper.thenOpen('http://localhost:3000/')
.then(function() {
  phantomcss.screenshot('#todo-app', 'Main app');
})
.then(function() {
  casper.fill('form.todo-form', {
    todo: 'Item1'
  }, true);

  phantomcss.screenshot('#todo-app', 'Item added');
});
```

File 2 - file2.js
```javascript
casper.thenOpen('http://localhost:3000/contact')
.then(function() {
  phantomcss.screenshot('#todo-app-contact', 'Main app');
})
.then(function() {
  casper.fill('form.todo-form-contact', {
    todo: 'Item2'
  }, true);

  phantomcss.screenshot('#todo-app-contact', 'Item added');
});
```

Subsequent files should call ```casper.then``` to continue the previous test.

```javascript
casper.then(function() {
  casper.click('.todo-done');

  phantomcss.screenshot('#todo-app', 'Item checked off');
});
```
You can also use ```casper.thenOpen``` to load a new url and continue testing


See the [CasperJS documentation](http://casperjs.readthedocs.org/en/latest/modules/casper.html) and the [PhantomCSS documentation](https://github.com/Huddle/PhantomCSS) for more information on using CasperJS and PhantomCSS.

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code using [Grunt](http://gruntjs.com/).

## Release History
* 2014-02-23   v0.2.8   Update dependencies
* 2014-02-23   v0.2.7   Remove bower, add dependencies node_modules, fix warning msgs
* 2014-02-23   v0.2.2   Added multiple file example to README.md
* 2014-02-07   v0.2.1   Fixed ResembleJS path issue
* 2014-01-07   v0.2.0   Merged updates from Larry Davis
* 2013-10-24   v0.1.1   Added the ability to use an external server
* 2013-10-24   v0.1.0   Initial Release


CHANGELOG.md
# CHANGELOG

### HEAD

### v0.4.1
* Added phantomjs as a global dependency.
* Updated docs

### v0.4.0
* Removed peerDependencies
* Updated phantomcss to v0.9.0
* Using casper from phantomcss
* No need for bower anymore

### v0.3.2
* Update dependencies
* Add `mismatchTolerance`as option
* Add `waitTimeout` as option

### v0.3.1
* Keep log level at "notice"
* Add time-grunt to measure build times
* Add jit-grunt to load tasks only when needed (+ improve performance)
* Change log level to notice in Gruntfile
* Document correct CasperJS log levels and pass them as arguments / option
* Added .editorconfig

### v0.3.0
* Added `mismatchTolerance` tolerance (`mismatchTolerance: 0.0.5` is default) (thx [fidgety](https://github.com/chrisgladd/grunt-phantomcss/pull/17))
* Use PhantomCSS v0.7.1
* Update dependencies
* Remove grunt as peerDependency causing multiple issues with peerPackages
* Fix post-script `bower install`


fixtures/coffeemachine_test.js
// Prevent error: "CasperError: casper.test property is only available using the `casperjs test` command"
phantom.casperTest = true;

casper.start('fixtures/coffeemachine.html').then(function () {
    phantomcss.screenshot('#coffee-machine-wrapper', 'open coffee machine button');
}).then( function () {
    casper.click('#coffee-machine-button');

    casper.waitForSelector('#myModal:not([style*="display: none"])',
        function success () {
            phantomcss.screenshot('#myModal', 'coffee machine dialog');
        },
        function timeout () {
            casper.test.fail('Should see coffee machine');
        }
    );
}).then( function () {
    casper.click('#cappuccino-button');
    phantomcss.screenshot('#myModal', 'cappuccino success');
}).then( function () {
    casper.click('#close');

    // wait for modal to fade-out
    casper.waitForSelector('#myModal[style*="display: none"]',
        function success () {
            phantomcss.screenshot('#coffee-machine-wrapper', 'coffee machine close success');
        },
        function timeout () {
            casper.test.fail('Should be able to walk away from the coffee machine');
        }
    );
});


phantomjs/runner.js
/*
* grunt-phantomcss
* https://github.com/anselmh/grunt-phantomcss
*
* Copyright (c) 2013 Chris Gladd
* Copyright (c) since 2014 Anselm Hannemann
*
* Licensed under the MIT license.
*/

'use-strict';

// Get node fileSystem module and define the separator module
var fs = require('fs');
var s = fs.separator;

// Parse arguments passed in from the grunt task
var args = JSON.parse(phantom.args[0]);

// Get viewport arguments (width | height)
var viewportSize = {
  width: args.viewportSize[0],
  height: args.viewportSize[1]
};

// Messages are sent to the parent by appending them to the tempfile
var sendMessage = function() {
  fs.write(args.tempFile, JSON.stringify(Array.prototype.slice.call(arguments)) + '\n', 'a');
};

// Initialise CasperJs
var phantomCSSPath = args.phantomCSSPath;

phantom.casperPath = args.casperPath;
phantom.injectJs(phantom.casperPath + s + 'bin' + s + 'bootstrap.js');

var casper = require('casper').create({
  viewportSize: viewportSize,
  logLevel: args.logLevel,
  verbose: true
});

// Require and initialise PhantomCSS module
var phantomcss = require(phantomCSSPath + s + 'phantomcss.js');

phantom.casperTest = true;

phantomcss.baseURL = args.baseURL;

phantomcss.init({
  screenshotRoot: args.screenshots,
  failedComparisonsRoot: args.failures,
  libraryRoot: phantomCSSPath, // Give absolute path, otherwise PhantomCSS fails
  mismatchTolerance: args.mismatchTolerance, // defaults to 0.05

  onFail: function(test) {
    sendMessage('onFail', test);
  },
  onPass: function(test) {
    sendMessage('onPass', test);
  },
  onTimeout: function(test) {
    sendMessage('onTimeout', test);
  },
  onComplete: function(allTests, noOfFails, noOfErrors) {
    sendMessage('onComplete', allTests, noOfFails, noOfErrors);
  }
});

casper.start();

// Run the test scenarios
args.test.forEach(function(testSuite) {
  require(testSuite);
});

// End tests and compare screenshots
casper.then(function() {
  phantomcss.compareAll();
})
.then(function() {
  casper.test.done();
})
.run(function() {
  phantom.exit();
});


tasks/phantomcss.js
/*
 * grunt-phantomcss
 * https://github.com/anselmh/grunt-phantomcss
 *
 * Copyright (c) 2013 Chris Gladd
 * Copyright (c) since 2014 Anselm Hannemann
 *
 * Licensed under the MIT license.
 */

'use strict';

var path = require('path');
var tmp = require('temporary');
var phantomBinaryPath = require('phantomjs').path;

var casperPath = path.join(__dirname, '..', 'node_modules', 'casperjs');

var runnerPath = path.join(__dirname, '..', 'phantomjs', 'runner.js');
var phantomCSSPath = path.join(__dirname, '..', 'node_modules', 'phantomcss');

module.exports = function(grunt) {
    grunt.registerMultiTask('phantomcss', 'CSS Regression Testing', function() {
        var done = this.async();

        // Variable object to set default values for options
        var options = this.options({
            baseURL: '',
            screenshots: 'screenshots',
            results: 'results',
            viewportSize: [1280, 800],
            mismatchTolerance: 0.05,
            waitTimeout: 5000, // Set timeout to wait before throwing an exception
            logLevel: 'warning' // debug | info | warning | error
        });

        // Timeout ID for message checking loop
        var messageCheckTimeout;

        // The number of tempfile lines already read
        var lastLine = 0;

        // The number of failed tests
        var failureCount = 0;

        // This is effectively the project root (location of Gruntfile)
        // This allows relative paths in tests, i.e. casper.start('someLocalFile.html')
        var cwd = process.cwd();

        // Create a temporary file for message passing between the task and PhantomJS
        var tempFile = new tmp.File();

        var deleteDiffScreenshots = function() {
            // Find diff/fail files
            var diffScreenshots = grunt.file.expand([
                path.join(options.screenshots, '*diff.png'),
                path.join(options.screenshots, '*fail.png')
            ]);

            // Delete all of 'em
            diffScreenshots.forEach(function(filepath) {
                grunt.file.delete(filepath);
            });
        };

        var cleanup = function(error) {
            // Remove temporary file
            tempFile.unlink();

            // Create the output directory
            grunt.file.mkdir(options.results);

            // Copy fixtures, diffs, and failure images to the results directory
            var allScreenshots = grunt.file.expand(path.join(options.screenshots, '**.png'));
            allScreenshots.forEach(function(filepath) {
                grunt.file.copy(filepath, path.join(options.results, path.basename(filepath)));
            });

            deleteDiffScreenshots();

            done(error || failureCount === 0);
        };

        var checkForMessages = function checkForMessages(stopChecking) {
            // Disable logging temporarily
            grunt.log.muted = true;

            // Read the file, splitting lines on \n, and removing a trailing line
            var lines = grunt.file.read(tempFile.path).split('\n').slice(0, -1);

            // Re-enable logging
            grunt.log.muted = false;

            // Iterate over all lines that haven't already been processed
            lines.slice(lastLine).some(function(line) {
                // Get args and method
                var args = JSON.parse(line);
                var eventName = args[0];

                // Debugging messages
                grunt.log.debug(JSON.stringify(['phantomjs'].concat(args)).magenta);

                // Call handler
                if (messageHandlers[eventName]) {
                    messageHandlers[eventName].apply(null, args.slice(1));
                }
            });

            // Update lastLine so previously processed lines are ignored
            lastLine = lines.length;

            if (stopChecking) {
                clearTimeout(messageCheckTimeout);
            }
            else {
                // Check back in a little bit
                messageCheckTimeout = setTimeout(checkForMessages, 100);
            }
        };

        var messageHandlers = {
            onFail: function(test) {
                grunt.log.writeln('Visual change found for ' + path.basename(test.filename) + ' (' + test.mismatch + '% mismatch)');
            },
            onPass: function(test) {
                grunt.log.writeln('No changes found for ' + path.basename(test.filename));
            },
            onTimeout: function(test) {
                grunt.log.writeln('Timeout while processing ' + path.basename(test.filename));
            },
            onComplete: function(allTests, noOfFails, noOfErrors) {
                if (allTests.length) {
                    var noOfPasses = allTests.length - failureCount;
                    failureCount = noOfFails + noOfErrors;

                    if (failureCount === 0) {
                        grunt.log.ok('All ' + noOfPasses + ' tests passed!');
                    }
                    else {
                        if (noOfErrors === 0) {
                            grunt.log.error(noOfFails + ' tests failed.');
                        }
                        else {
                            grunt.log.error(noOfFails + ' tests failed, ' + noOfErrors + ' had errors.');
                        }
                    }
                }
                else {
                    grunt.log.ok('Baseline screenshots generated in '+options.screenshots);
                    grunt.log.warn('Check that the generated screenshots are visually correct and delete them if they aren\'t.');
                }
            }
        };

        // Resolve paths for tests
        options.test = [];
        this.filesSrc.forEach(function(filepath) {
            options.test.push(path.resolve(filepath));
        });

        options.screenshots = path.resolve(options.screenshots);

        // Put failure screenshots in the same place as source screenshots, we'll move/delete them after the test run
        // Note: This duplicate assignment is provided for clarity; PhantomCSS will put failures in the screenshots folder by default
        options.failures = options.screenshots;

        // Pass necessary paths
        options.tempFile = tempFile.path;
        options.phantomCSSPath = phantomCSSPath;
        options.casperPath = casperPath;


        // Remove old diff screenshots
        deleteDiffScreenshots();

        // Start watching for messages
        checkForMessages();

        grunt.util.spawn({
            cmd: phantomBinaryPath,
            args: [
                runnerPath,
                JSON.stringify(options)
            ],
            opts: {
                cwd: cwd,
                stdio: 'inherit'
            }
        }, function(error, result, code) {
            // When Phantom exits check for remaining messages one last time
            checkForMessages(true);

            cleanup(error);
        });
    });
};
